File: RestClientApp/src/com/restclient/client/widgets/JSONViewer.java
package com.restclient.client.widgets;

import java.util.Iterator;
import java.util.Set;

import com.google.gwt.core.client.GWT;
import com.google.gwt.json.client.JSONArray;
import com.google.gwt.json.client.JSONBoolean;
import com.google.gwt.json.client.JSONNull;
import com.google.gwt.json.client.JSONNumber;
import com.google.gwt.json.client.JSONObject;
import com.google.gwt.json.client.JSONParser;
import com.google.gwt.json.client.JSONString;
import com.google.gwt.json.client.JSONValue;
import com.google.gwt.resources.client.CssResource;
import com.google.gwt.safehtml.shared.SafeHtmlUtils;
import com.google.gwt.uibinder.client.UiBinder;
import com.google.gwt.uibinder.client.UiField;
import com.google.gwt.user.client.ui.Composite;
import com.google.gwt.user.client.ui.HTML;
import com.google.gwt.user.client.ui.HTMLPanel;
import com.google.gwt.user.client.ui.Widget;

public class JSONViewer extends Composite {
	
	interface Binder extends UiBinder<Widget, JSONViewer> {}
	
	interface ParserStyle extends CssResource {
		int indent();
		String prettyPrint();
		String numeric();
		String nullValue();
		String booleanValue();
		String punctuation();
		String stringValue();
		String node();
		String arrayCounter();
		String keyName();
	}
	
	@UiField HTML result;
	@UiField ParserStyle style;
	
	
	JSONValue jsonValue = null;
	
	public JSONViewer(String data, HTMLPanel jsonPanel) {
		
		initWidget(GWT.<Binder> create(Binder.class).createAndBindUi(this));
		
		try {
			this.jsonValue = JSONParser.parseLenient(data);
			
		} catch (NullPointerException e) {
			
		} catch (IllegalArgumentException e) {
			
		} finally {
			
			if( this.jsonValue == null ){
				//TODO - set parse error
			} else {
				String html = this.parse( this.jsonValue, 1 );
				String parsedData = "<div class=\""+style.prettyPrint()+"\">";
				parsedData += html;
				parsedData += "</div>";
				result.setHTML(parsedData);
				jsonPanel.add(this);
			}
		}
	}
	
	private String parse(JSONValue data, int indent){
		String result = "";
		JSONObject obj = data.isObject(); //JSON Object or null
		JSONArray arr = data.isArray(); //JSON array or null
		JSONString str = data.isString();//JSON string or null
		JSONBoolean bool = data.isBoolean();//JSON boolean or null
		JSONNull isNull = data.isNull(); //JSON null or null
		JSONNumber number = data.isNumber();//JSON number ornull
		
		if( obj != null ){
			result += this.parseObject(obj, indent);
		} else if( arr != null ) {
			result += this.parseArray(arr, indent);
		} else if( str != null ) {
			result += this.parseStringValue(str);
		} else if( bool != null ) {
			result += this.parseBooleanValue(bool);
		} else if( isNull != null ) {
			result += this.parseNullValue(isNull);
		} else if( number != null ) {
			result += this.parseNumericValue(number);
		}
		
		return result;
	}
	
	private String parseNumericValue(JSONNumber number) {
		String result = "";
		result += "<span class=\""+style.numeric()+"\">";
		result += number.toString();
		result += "</span>";
		return result;
	}

	private String parseNullValue(JSONNull isNull) {
		String result = "";
		result += "<span class=\""+style.nullValue()+"\">";
		result += "null";
		result += "</span>";
		return result;
	}

	private String parseBooleanValue(JSONBoolean bool) {
		String result = "";
		result += "<span class=\""+style.booleanValue()+"\">";
		if( bool != null )
			result += String.valueOf(bool.booleanValue());
		else 
			result += "null";
		result += "</span>";
		return result;
	}

	private String parseStringValue(JSONString str) {
		String result = "";
		result += "<span class=\""+style.punctuation()+"\">&quot;</span>";
		result += "<span class=\""+style.stringValue()+"\">";
		if( str != null )
			result += SafeHtmlUtils.htmlEscape(str.stringValue());
		else
			result += "null";
		result += "</span>";
		result += "<span class=\""+style.punctuation()+"\">&quot;</span>";
		return result;
	}

	private String parseObject(JSONObject object, int indent) {
		String result = "";
		
		result += "<span class=\""+style.punctuation()+"\">{</span>";
		
		Set<String> keys = object.keySet();
		Iterator<String> it = keys.iterator();
		
		while( it.hasNext() ){
			int indentValue = indent*style.indent();
			
			result += "<div style=\"text-indent:"+indentValue+"px\" class=\""+style.node()+"\">";
			String key = it.next();
			JSONValue value = object.get(key);
			String data = this.parse(value, indent+1);
			result += this.parseKey(key) +": "+data;
			if(it.hasNext()){
				result += "<span class=\""+style.punctuation()+"\">,</span>";
			}
			result += "</div>";
		}
		
		
		result += "<span class=\""+style.punctuation()+"\">}</span>";
		return result;
	}
	
	
	
	
	private String parseArray(JSONArray array, int indent) {
		String result = "";
		
		result += "<span class=\""+style.punctuation()+"\">[</span>";
		int cnt = array.size();
		
		result += "<span class=\""+style.arrayCounter()+"\">("+cnt+")</span>";
		
		for( int i = 0; i<cnt; i++ ){
			int indentValue = indent*style.indent();
			
			result += "<div style=\"text-indent:"+indentValue+"px\" class=\""+style.node()+"\">";
			JSONValue value = array.get(i);
			result += this.parse( value, indent+1 );
			if( i<cnt-1 ){
				result += "<span class=\""+style.punctuation()+"\">,</span>";
			}
			result += "</div>";
		}
		
		result += "<span class=\""+style.punctuation()+"\">]</span>";
		return result;
	}
	
	private String parseKey( String key ){
		String result = "";
		result += "<span class=\""+style.punctuation()+"\">&quot;</span>";
		result += "<span class=\""+style.keyName()+"\">"+key+"</span>";
		result += "<span class=\""+style.punctuation()+"\">&quot;</span>";
		return result;
	}
	
}

File: RestClient/war/workers/jsonviewer.js
self.onmessage = function(e) {
	
    var data = JSON.parse(e.data);
    //expected data: {style: {style definitions}, data: data to parse}
    var style = data.style;
    for(var styleName in style){
    	JSONViewer.STYLE[styleName] = style[styleName]; 
    }
    
    var parser = new JSONViewer(data.data);
    if(parser.latestError!=null){
        self.postMessage(parser.latestError);
        return;
    }
    
    var result = parser.getHTML();
    parser = null;
    self.postMessage(result);
};

function JSONViewer(data){
	this.linkRegExp = new RegExp(/^\w+:\/\//gim);
    this.jsonValue = null;
    this.latestError = null;
    this.elementsCounter = 0;
    if(typeof data == "string"){
        try {
            this.jsonValue = JSON.parse(data);
        } catch (e) {
            this.latestError = e.message;
        }
    } else {
        this.jsonValue = data;
    }
}

JSONViewer.STYLE = {
	prettyPrint: "prettyPrint",
	numeric: "numeric",
	nullValue: "nullValue",
	booleanValue: "booleanValue",
	punctuation: "punctuation",
	stringValue: "stringValue",
	node: "node",
	arrayCounter: "arrayCounter",
	keyName: "keyName",
	rootElementToggleButton: "rootElementToggleButton",
	infoRow: "infoRow",
	brace: "brace"
};

JSONViewer.prototype = {
    getHTML: function(){
    	var parsedData = '<div class="'+JSONViewer.STYLE.prettyPrint+'">';
        parsedData += this.parse(this.jsonValue);
        parsedData += "</div>";
        return parsedData;
    },
    /**
     * Parse JSON data
     */
    parse: function(data){
    	
		var result = "";
		if(data == null){
			result += this.parseNullValue();
		} else if(typeof data == "number"){
			result += this.parseNumericValue(data);
		} else if(typeof data == "boolean"){
			result += this.parseBooleanValue(data);
		} else if(typeof data == "string"){
			result += this.parseStringValue(data);
		} else if(data instanceof Array){
			result += this.parseArray(data);
		} else {
			result += this.parseObject(data);
		}
		
		return result;
    },
    
    parseNullValue: function(){
    	var result = "";
		result += '<span class="'+JSONViewer.STYLE.nullValue+'">';
		result += "null";
		result += '</span>';
		return result;
    },
    
    parseNumericValue: function(number) {
    	var result = "";
		result += '<span class="'+JSONViewer.STYLE.numeric+'">';
		result += number+"";
		result += '</span>';
		return result;
	},
    
    parseBooleanValue: function(bool) {
    	var result = "";
		result += '<span class="'+JSONViewer.STYLE.booleanValue+'">';
		if(bool != null)
			result += bool+"";
		else 
			result += "null";
		result += "</span>";
		return result;
	},
	
	parseStringValue: function(str) {
		var result = "";
		var value = str || "";
		if(value != null){
			value = SafeHtmlUtils.htmlEscape(value);
			try{
				if(this.linkRegExp.test(value)){
					value = '<a response-anchor href="'+value+'">'+value+'</a>';
				}
			} catch(e){}
		} else {
			value = "null";
		}
		result += '<span class="'+JSONViewer.STYLE.punctuation+'">&quot;</span>';
		result += '<span class="'+JSONViewer.STYLE.stringValue+'">';
		result += value;
		result += '</span>';
		result += '<span class="'+JSONViewer.STYLE.punctuation+'">&quot;</span>';
		return result;
	},
	
	parseObject: function(object) {
		var result = "";
		result += '<div class="'+JSONViewer.STYLE.punctuation + " " + JSONViewer.STYLE.brace + '">{</div>';
		result += '<div collapse-indicator class="'+JSONViewer.STYLE.infoRow+'">...</div>';
		var pairs = [];
		for(var key in object){
			if(!object.hasOwnProperty(key)) continue;
			pairs[pairs.length] = key;
			pairs[pairs.length] = object[key];
		}
		var cnt = pairs.length;
		for(var i=0; i<cnt; i=i+2){
			var key = pairs[i];
			var value = pairs[i+1];
			var elementNo = this.elementsCounter++;
			var data = this.parse(value);
			var hasManyChildren = this.elementsCounter - elementNo > 1;
			
			result += '<div data-element="'+elementNo+'" style="margin-left: 15px" class="'+JSONViewer.STYLE.node+'">';
			result += this.parseKey(key) + ": " + data;
			if(i+2 != cnt){
				result += '<span class="'+JSONViewer.STYLE.punctuation+'">,</span>';
			}
			if(hasManyChildren){
				result += '<div data-toggle="'+elementNo+'" class="'+JSONViewer.STYLE.rootElementToggleButton+'">-</div>';
			}
			result += "</div>";
		}
		
		result += '<div class="' + JSONViewer.STYLE.punctuation + " " + JSONViewer.STYLE.brace + '">}</div>';
		return result;
	},
	
	parseArray: function(array) {
		var result = "";
		result += '<div class="'+JSONViewer.STYLE.punctuation + " " + JSONViewer.STYLE.brace + '">[</div>';
		result += '<div collapse-indicator class="'+JSONViewer.STYLE.infoRow+'">...</div>';
		var cnt = array.length;
		result += '<span class="'+JSONViewer.STYLE.arrayCounter+'">('+cnt+')</span>';
		
		for(var i = 0; i<cnt; i++){
			var elementNo = this.elementsCounter++;
			var data = this.parse(array[i]);
			var hasManyChildren = this.elementsCounter - elementNo > 1;
			
			result += '<div data-element="'+elementNo+'" style="margin-left: 15px" class="'+JSONViewer.STYLE.node+'">';
			result += data;
			if(i<cnt-1){
				result += '<span class="'+JSONViewer.STYLE.punctuation+'">,</span>';
			}
			if(hasManyChildren){
				result += '<div data-toggle="'+elementNo+'" class="'+JSONViewer.STYLE.rootElementToggleButton+'">-</div>';
			}
			result += "</div>";
		}
		
		result += "<span class=\""+JSONViewer.STYLE.punctuation +  " " + JSONViewer.STYLE.brace + "\">]</span>";
		return result;
	},
	
	parseKey: function(key){
		var result = "";
		result += '<span class="'+JSONViewer.STYLE.punctuation+'">&quot;</span>';
		result += '<span class="'+JSONViewer.STYLE.keyName+'">'+key+'</span>';
		result += '<span class="'+JSONViewer.STYLE.punctuation+'">&quot;</span>';
		return result;
	}
}

var SafeHtmlUtils = {
		AMP_RE: new RegExp(/&/g),
		GT_RE: new RegExp(/>/g),
		LT_RE: new RegExp(/</g),
		SQUOT_RE: new RegExp(/'/g),
		QUOT_RE: new RegExp(/"/g),

		htmlEscape: function(s){
			if (s.indexOf("&") != -1) {
		      s = s.replace(SafeHtmlUtils.AMP_RE, '&amp;')
		    }
		    if (s.indexOf("<") != -1) {
		    	s = s.replace(SafeHtmlUtils.LT_RE, '&lt;')
		    }
		    if (s.indexOf(">") != -1) {
		    	s = s.replace(SafeHtmlUtils.GT_RE, '&gt;')
		    }
		    if (s.indexOf("\"") != -1) {
		    	s = s.replace(SafeHtmlUtils.QUOT_RE, '&quot;')
		    }
		    if (s.indexOf("'") != -1) {
		    	s = s.replace(SafeHtmlUtils.SQUOT_RE, '&#39;')
		    }
		    return s;
		}	
}